README.md
# T4 Support

Very basic language support for T4 Text Templates (.tt, .ttinclude) files.

For more information about using T4 templates, see [Code Generation and T4 Text Templates](https://docs.microsoft.com/en-us/visualstudio/modeling/code-generation-and-t4-text-templates).

## Features

Provides `C#` syntax highlighting for embedded code within T4 templates as well as snippets for common directives and functions.

![Light theme screenshot](./assets/screenshot1.png)

![Dark theme screenshot](./assets/screenshot2.png)

## Known Issues

Currently only `C#` embedded highlighting is supported.

## Release Notes

### 0.5.0

Recognize `.ttinclude` files.

### 0.4.4

Fix inverted highlighting on back-to-back blocks.

### 0.4.3

Fixed compatibility for files without `.tt` extension. Thank you @awglyde!

### 0.4.0

Add background highlighting of open/closing brackets (<#, #>) and code blocks for better visual separation.

### 0.3.2

Further fixes for syntax highlighting.

### 0.3.1

Minor fix in syntax highlighting.

### 0.3.0

Added snippets!

### 0.2.2

Spruced up inline code regex, which fixed an issue where code failed to exit a block.

### 0.2.1

Fix issue where curly braces inside code blocks prevented escaping from block.

### 0.2.0

Rudimentary highlighting for opening/closing brackets.

### 0.1.3

Add extension icon.

### 0.1.2

Fixed sloppy highlighting.

### 0.1.1

Support for VS Code versions 1.0.0 and above.

### 0.1.0

Initial release.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.7.0] - 2023-03-09

- Add "Go To File" support for included files. Thank you [Jeff](https://github.com/sea1jxr)!

## [0.6.0] - 2022-08-26

- Recognize `.t4` files.  Thank you [Brice](https://github.com/bricelam)!

## [0.5.0] - 2020-05-22

- Recognize `.ttinclude` files.

## [0.4.4] - 2020-05-14

- Fix inverted highlighting on back-to-back blocks.

## [0.4.3] - 2019-04-15

- Fixed compatibility for files without `.tt` extension. Thank you @awglyde!

## [0.4.2] - 2018-03-19

- Update icon to match vscode-icons.

## [0.4.1] - 2018-03-08

- Docs update.

## [0.4.0] - 2018-02-27

- Add background highlighting of open/closing brackets (<#, #>) and code blocks for better visual separation.

## [0.3.2] - 2018-02-26

- Further fixes for syntax highlighting.

## [0.3.1] - 2018-02-22

- Minor fix in syntax highlighting.

## [0.3.0] - 2018-02-21

- Added some snippets
- More refinement of syntax highlighting.

## [0.2.2] - 2018-02-21

- Fixed an issue where inline code failed to exit code block.

## [0.2.1] - 2018-02-21

- Fix issue where curly braces inside code blocks prevented escaping from block.

## [0.2.0] - 2018-02-21

- Highlighting for open/closing brackets.

## [0.1.3] - 2018-02-21

- Add extension icon.

## [0.1.2] - 2018-02-21

- Fixed sloppy highlighting.

## [0.1.1] - 2018-02-21

- Support for VS Code versions 1.0.0 and above.

## [0.1.0] - 2018-02-21

- Initial release with basic embedded syntax highlighting.


src/extension.ts
import * as fs from 'fs';
import * as path from 'path';
import * as vscode from 'vscode';

export function activate(context: vscode.ExtensionContext) {

	context.subscriptions.push(vscode.languages.registerDefinitionProvider({ language: 't4' }, new T4DefinitionProvider()));
	const bracketDecorationType = vscode.window.createTextEditorDecorationType({
		light: {
			backgroundColor: 'rgba(255, 100, 0, .2)'
		},
		dark: {
			backgroundColor: 'rgba(0, 100, 255, .2)'
		}
	});

	const codeBlockDecorationType = vscode.window.createTextEditorDecorationType({
		light: {
			backgroundColor: 'rgba(100,100,100,0.1)'
		},
		dark: {
			backgroundColor: 'rgba(220,220,220,0.1)'
		}
	});

	let activeEditor = vscode.window.activeTextEditor;
	if (activeEditor) {
		triggerUpdateDecorations();
	}

	vscode.window.onDidChangeActiveTextEditor(editor => {
		activeEditor = editor;
		if (editor) {
			triggerUpdateDecorations();
		}
	}, null, context.subscriptions);

	vscode.workspace.onDidChangeTextDocument(event => {
		if (activeEditor && event.document === activeEditor.document) {
			triggerUpdateDecorations();
		}
	}, null, context.subscriptions);

	var timeout = null;
	function triggerUpdateDecorations() {
		if (timeout) {
			clearTimeout(timeout);
		}
		timeout = setTimeout(updateDecorations, 200);
	}

	function updateDecorations() {
		if (!activeEditor || !isT4File(activeEditor)) {
			return;
		}
		const regEx = /(<#@|<#\+|<#=|<#)|(#>)+/g;
		const text = activeEditor.document.getText();
		const brackets: vscode.DecorationOptions[] = [];
		let match: RegExpExecArray;
		while (match = regEx.exec(text)) {
			const startPos = activeEditor.document.positionAt(match.index);
			const endPos = activeEditor.document.positionAt(match.index + match[0].length);
			const decoration = { range: new vscode.Range(startPos, endPos), hoverMessage: 'T4 Bracket **' + match[0] + '**' };

			brackets.push(decoration);
		}


		const blocks: vscode.DecorationOptions[] = [];

		let index = 0;
		let max = brackets.length;
		brackets.forEach(element => {

			if(index + 1 < max) {
				const start = brackets[index];
				const end = brackets[index + 1];

				const decoration = { range: new vscode.Range(start.range.end, end.range.start), hoverMessage: "" };

				blocks.push(decoration);
			}

			index += 2;
		});

		activeEditor.setDecorations(bracketDecorationType, brackets);
		activeEditor.setDecorations(codeBlockDecorationType, blocks);
	}

	function isT4File(editor: vscode.TextEditor): boolean{
		return editor.document.languageId == "t4";
	}
}

class T4DefinitionProvider implements vscode.DefinitionProvider {
  public provideDefinition(
    document: vscode.TextDocument,
    position: vscode.Position,
    token: vscode.CancellationToken
  ): vscode.ProviderResult<vscode.Definition | vscode.DefinitionLink[]> {
    return new Promise((resolve, reject) => {
      const line = document.lineAt(position.line).text;
      const regEx = /<#@\s*include\s*file\s*=\s*"(?<filepath>[^"]*)"\s*#>/dg;

      const match: RegExpExecArray = regEx.exec(line);
      const originStart = (match as any)?.indices.groups.filepath[0];
      const originEnd = (match as any)?.indices.groups.filepath[1];
      if (!match || position.character < originStart || position.character > originEnd) {
        reject();
        return;
      }

      let definitionFilepath = (match as any).groups.filepath;
      if (!path.isAbsolute(definitionFilepath)) {
        const dirName = path.dirname(document.fileName);
        definitionFilepath = path.join(dirName, definitionFilepath);
      }

      if (!fs.existsSync(definitionFilepath)) {
        reject();
        return;
      }

      resolve([
        {
          originSelectionRange: new vscode.Range(
            position.with({ character: originStart }),
            position.with({ character: originEnd })
          ),
          targetUri: vscode.Uri.file(definitionFilepath),
          targetRange: new vscode.Range(
            new vscode.Position(0, 0),
            new vscode.Position(0, 0)
          ),
        },
      ]);
    });
  }
}


out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activate = void 0;
var fs = require("fs");
var path = require("path");
var vscode = require("vscode");
function activate(context) {
    context.subscriptions.push(vscode.languages.registerDefinitionProvider({ language: 't4' }, new T4DefinitionProvider()));
    var bracketDecorationType = vscode.window.createTextEditorDecorationType({
        light: {
            backgroundColor: 'rgba(255, 100, 0, .2)'
        },
        dark: {
            backgroundColor: 'rgba(0, 100, 255, .2)'
        }
    });
    var codeBlockDecorationType = vscode.window.createTextEditorDecorationType({
        light: {
            backgroundColor: 'rgba(100,100,100,0.1)'
        },
        dark: {
            backgroundColor: 'rgba(220,220,220,0.1)'
        }
    });
    var activeEditor = vscode.window.activeTextEditor;
    if (activeEditor) {
        triggerUpdateDecorations();
    }
    vscode.window.onDidChangeActiveTextEditor(function (editor) {
        activeEditor = editor;
        if (editor) {
            triggerUpdateDecorations();
        }
    }, null, context.subscriptions);
    vscode.workspace.onDidChangeTextDocument(function (event) {
        if (activeEditor && event.document === activeEditor.document) {
            triggerUpdateDecorations();
        }
    }, null, context.subscriptions);
    var timeout = null;
    function triggerUpdateDecorations() {
        if (timeout) {
            clearTimeout(timeout);
        }
        timeout = setTimeout(updateDecorations, 200);
    }
    function updateDecorations() {
        if (!activeEditor || !isT4File(activeEditor)) {
            return;
        }
        var regEx = /(<#@|<#\+|<#=|<#)|(#>)+/g;
        var text = activeEditor.document.getText();
        var brackets = [];
        var match;
        while (match = regEx.exec(text)) {
            var startPos = activeEditor.document.positionAt(match.index);
            var endPos = activeEditor.document.positionAt(match.index + match[0].length);
            var decoration = { range: new vscode.Range(startPos, endPos), hoverMessage: 'T4 Bracket **' + match[0] + '**' };
            brackets.push(decoration);
        }
        var blocks = [];
        var index = 0;
        var max = brackets.length;
        brackets.forEach(function (element) {
            if (index + 1 < max) {
                var start = brackets[index];
                var end = brackets[index + 1];
                var decoration = { range: new vscode.Range(start.range.end, end.range.start), hoverMessage: "" };
                blocks.push(decoration);
            }
            index += 2;
        });
        activeEditor.setDecorations(bracketDecorationType, brackets);
        activeEditor.setDecorations(codeBlockDecorationType, blocks);
    }
    function isT4File(editor) {
        return editor.document.languageId == "t4";
    }
}
exports.activate = activate;
var T4DefinitionProvider = /** @class */ (function () {
    function T4DefinitionProvider() {
    }
    T4DefinitionProvider.prototype.provideDefinition = function (document, position, token) {
        return new Promise(function (resolve, reject) {
            var line = document.lineAt(position.line).text;
            var regEx = /<#@\s*include\s*file\s*=\s*"(?<filepath>[^"]*)"\s*#>/dg;
            var match = regEx.exec(line);
            var originStart = match === null || match === void 0 ? void 0 : match.indices.groups.filepath[0];
            var originEnd = match === null || match === void 0 ? void 0 : match.indices.groups.filepath[1];
            if (!match || position.character < originStart || position.character > originEnd) {
                reject();
                return;
            }
            var definitionFilepath = match.groups.filepath;
            if (!path.isAbsolute(definitionFilepath)) {
                var dirName = path.dirname(document.fileName);
                definitionFilepath = path.join(dirName, definitionFilepath);
            }
            if (!fs.existsSync(definitionFilepath)) {
                reject();
                return;
            }
            resolve([
                {
                    originSelectionRange: new vscode.Range(position.with({ character: originStart }), position.with({ character: originEnd })),
                    targetUri: vscode.Uri.file(definitionFilepath),
                    targetRange: new vscode.Range(new vscode.Position(0, 0), new vscode.Position(0, 0)),
                },
            ]);
        });
    };
    return T4DefinitionProvider;
}());
//# sourceMappingURL=extension.js.map